In [59]:
# import libraries
import pandas as pd
import numpy as np

In [60]:
pd.set_option('display.max_columns', None)

### Load dataset

In [61]:
# load csv file
df = pd.read_csv("../data/HR-Employee-Attrition.csv")
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0  41       Yes      Travel_Rarely     1102.0                   Sales   
1  49        No  Travel_Frequently      279.0  Research & Development   
2  37       Yes      Travel_Rarely     1373.0  Research & Development   
3  33        No  Travel_Frequently     1392.0  Research & Development   
4  27        No      Travel_Rarely      591.0  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0               1.0        2.0  Life Sciences            1.0             1.0   
1               8.0        1.0  Life Sciences            1.0             2.0   
2               2.0        2.0          Other            1.0             4.0   
3               3.0        4.0  Life Sciences            1.0             5.0   
4               2.0        1.0        Medical            1.0             7.0   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                      2.0  Female        94.0             3.0       2.0   
1                      3.0    Male        61.0             2.0       2.0   
2                      4.0    Male        92.0             2.0       1.0   
3                      4.0  Female        56.0             3.0       1.0   
4                      1.0    Male        40.0             3.0       1.0   

                 JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  \
0        Sales Executive              4.0        Single         5993.0   
1     Research Scientist              2.0       Married         5130.0   
2  Laboratory Technician              3.0        Single         2090.0   
3     Research Scientist              3.0       Married         2909.0   
4  Laboratory Technician              2.0       Married         3468.0   

   MonthlyRate  NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
0      19479.0                 8.0      Y      Yes               11.0   
1      24907.0                 1.0      Y       No               23.0   
2       2396.0                 6.0      Y      Yes               15.0   
3      23159.0                 1.0      Y      Yes               11.0   
4      16632.0                 9.0      Y       No               12.0   

   PerformanceRating  RelationshipSatisfaction  StandardHours  \
0                3.0                       1.0           80.0   
1                4.0                       4.0           80.0   
2                3.0                       2.0           80.0   
3                3.0                       3.0           80.0   
4                3.0                       4.0           80.0   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0               0.0                8.0                    0.0   
1               1.0               10.0                    3.0   
2               0.0                7.0                    3.0   
3               0.0                8.0                    3.0   
4               1.0                6.0                    3.0   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0              1.0             6.0                 4.0   
1              3.0            10.0                 7.0   
2              3.0             0.0                 0.0   
3              3.0             8.0                 7.0   
4              3.0             2.0                 2.0   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                      0.0                   5.0  
1                      1.0                   7.0  
2                      0.0                   0.0  
3                      3.0                   0.0  
4                      2.0                   2.0

### Initial Data Profiling

In [62]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1476 entries, 0 to 1475
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       1476 non-null   object 
 1   Attrition                 1476 non-null   object 
 2   BusinessTravel            1470 non-null   object 
 3   DailyRate                 1470 non-null   float64
 4   Department                1470 non-null   object 
 5   DistanceFromHome          1470 non-null   float64
 6   Education                 1470 non-null   float64
 7   EducationField            1470 non-null   object 
 8   EmployeeCount             1470 non-null   float64
 9   EmployeeNumber            1470 non-null   float64
 10  EnvironmentSatisfaction   1470 non-null   float64
 11  Gender                    1470 non-null   object 
 12  HourlyRate                1470 non-null   float64
 13  JobInvolvement            1470 non-null   float64
 14  JobLevel

In [63]:
print(f' Number of rows: {df.shape[0]}')
print(f' Number of columns: {df.shape[1]}')

 Number of rows: 1476
 Number of columns: 35


In [64]:
df.describe()

DailyRate  DistanceFromHome    Education  EmployeeCount  \
count  1470.000000       1470.000000  1470.000000         1470.0   
mean    802.485714          9.192517     2.912925            1.0   
std     403.509100          8.106864     1.024165            0.0   
min     102.000000          1.000000     1.000000            1.0   
25%     465.000000          2.000000     2.000000            1.0   
50%     802.000000          7.000000     3.000000            1.0   
75%    1157.000000         14.000000     4.000000            1.0   
max    1499.000000         29.000000     5.000000            1.0   

       EmployeeNumber  EnvironmentSatisfaction   HourlyRate  JobInvolvement  \
count     1470.000000              1470.000000  1470.000000     1470.000000   
mean      1024.865306                 2.721769    65.891156        2.729932   
std        602.024335                 1.093082    20.329428        0.711561   
min          1.000000                 1.000000    30.000000        1.000000   
25%        491.250000                 2.000000    48.000000        2.000000   
50%       1020.500000                 3.000000    66.000000        3.000000   
75%       1555.750000                 4.000000    83.750000        3.000000   
max       2068.000000                 4.000000   100.000000        4.000000   

          JobLevel  JobSatisfaction  MonthlyIncome   MonthlyRate  \
count  1470.000000      1470.000000    1470.000000   1470.000000   
mean      2.063946         2.728571    6502.931293  14313.103401   
std       1.106940         1.102846    4707.956783   7117.786044   
min       1.000000         1.000000    1009.000000   2094.000000   
25%       1.000000         2.000000    2911.000000   8047.000000   
50%       2.000000         3.000000    4919.000000  14235.500000   
75%       3.000000         4.000000    8379.000000  20461.500000   
max       5.000000         4.000000   19999.000000  26999.000000   

       NumCompaniesWorked  PercentSalaryHike  PerformanceRating  \
count         1470.000000        1470.000000        1470.000000   
mean             2.693197          15.209524           3.153741   
std              2.498009           3.659938           0.360824   
min              0.000000          11.000000           3.000000   
25%              1.000000          12.000000           3.000000   
50%              2.000000          14.000000           3.000000   
75%              4.000000          18.000000           3.000000   
max              9.000000          25.000000           4.000000   

       RelationshipSatisfaction  StandardHours  StockOptionLevel  \
count               1470.000000         1470.0       1470.000000   
mean                   2.712245           80.0          0.793878   
std                    1.081209            0.0          0.852077   
min                    1.000000           80.0          0.000000   
25%                    2.000000           80.0          0.000000   
50%                    3.000000           80.0          1.000000   
75%                    4.000000           80.0          1.000000   
max                    4.000000           80.0          3.000000   

       TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  \
count        1470.000000            1470.000000      1470.000000   
mean           11.279592               2.799320         2.761224   
std             7.780782               1.289271         0.706476   
min             0.000000               0.000000         1.000000   
25%             6.000000               2.000000         2.000000   
50%            10.000000               3.000000         3.000000   
75%            15.000000               3.000000         3.000000   
max            40.000000               6.000000         4.000000   

       YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
count     1470.000000         1470.000000              1470.000000   
mean         7.008163            4.229252                 2.187755   
std          6.126525          

There are 1476 employees with 35 variables, including target variable "Attrition."

### Missing Data
Check if there are missing values that need handling.

In [65]:
# Check what columns have missing values
missing = df.isnull().sum()
missing[missing >0]

BusinessTravel              6
DailyRate                   6
Department                  6
DistanceFromHome            6
Education                   6
EducationField              6
EmployeeCount               6
EmployeeNumber              6
EnvironmentSatisfaction     6
Gender                      6
HourlyRate                  6
JobInvolvement              6
JobLevel                    6
JobRole                     6
JobSatisfaction             6
MaritalStatus               6
MonthlyIncome               6
MonthlyRate                 6
NumCompaniesWorked          6
Over18                      6
OverTime                    6
PercentSalaryHike           6
PerformanceRating           6
RelationshipSatisfaction    6
StandardHours               6
StockOptionLevel            6
TotalWorkingYears           6
TrainingTimesLastYear       6
WorkLifeBalance             6
YearsAtCompany              6
YearsInCurrentRole          6
YearsSinceLastPromotion     6
YearsWithCurrManager        6
dtype: int

In [66]:
# Calculate percentage of values missing
missing_df = pd.DataFrame({
    "Missing_Count": df.isnull().sum(),
    "Percentage": (df.isnull().sum() / len(df)) * 100})
missing_df[missing_df['Missing_Count'] > 0]

Missing_Count  Percentage
BusinessTravel                        6    0.406504
DailyRate                             6    0.406504
Department                            6    0.406504
DistanceFromHome                      6    0.406504
Education                             6    0.406504
EducationField                        6    0.406504
EmployeeCount                         6    0.406504
EmployeeNumber                        6    0.406504
EnvironmentSatisfaction               6    0.406504
Gender                                6    0.406504
HourlyRate                            6    0.406504
JobInvolvement                        6    0.406504
JobLevel                              6    0.406504
JobRole                               6    0.406504
JobSatisfaction                       6    0.406504
MaritalStatus                         6    0.406504
MonthlyIncome                         6    0.406504
MonthlyRate                           6    0.406504
NumCompaniesWorked                    6    0.406504
Over18                                6    0.406504
OverTime                              6    0.406504
PercentSalaryHike                     6    0.406504
PerformanceRating                     6    0.406504
RelationshipSatisfaction              6    0.406504
StandardHours                         6    0.406504
StockOptionLevel                      6    0.406504
TotalWorkingYears                     6    0.406504
TrainingTimesLastYear                 6    0.406504
WorkLifeBalance                       6    0.406504
YearsAtCompany                        6    0.406504
YearsInCurrentRole                    6    0.406504
YearsSinceLastPromotion               6    0.406504
YearsWithCurrManager                  6    0.406504

In [67]:
# Number of columns with missing values
print(f"Number of columns with missing values: {len(missing_df[missing_df['Missing_Count'] > 0])}")

Number of columns with missing values: 33


In [68]:
# Find all rows with missing values
rows_with_missing = df[df.isnull().any(axis=1)]

print(f'Number of rows with missing values: {len(rows_with_missing)}')

Number of rows with missing values: 6


In [69]:
# Examine rows with missing values
rows_with_missing

Age Attrition  \
1470             # Updated documentation on 10 May 2024     10:07   
1471     # Improved visualization layout on 04 Jul 2024     15:16   
1472      # Enhanced prediction accuracy on 13 Jul 2024     12:15   
1473  # Improved feature engineering steps on 14 Jul...     14:15   
1474      # Enhanced prediction accuracy on 26 Jul 2024     10:09   
1475            # Cleaned up unused code on 28 Jul 2024     16:24   

     BusinessTravel  DailyRate Department  DistanceFromHome  Education  \
1470            NaN        NaN        NaN               NaN        NaN   
1471            NaN        NaN        NaN               NaN        NaN   
1472            NaN        NaN        NaN               NaN        NaN   
1473            NaN        NaN        NaN               NaN        NaN   
1474            NaN        NaN        NaN               NaN        NaN   
1475            NaN        NaN        NaN               NaN        NaN   

     EducationField  EmployeeCount  EmployeeNumber  EnvironmentSatisfaction  \
1470            NaN            NaN             NaN                      NaN   
1471            NaN            NaN             NaN                      NaN   
1472            NaN            NaN             NaN                      NaN   
1473            NaN            NaN             NaN                      NaN   
1474            NaN            NaN             NaN                      NaN   
1475            NaN            NaN             NaN                      NaN   

     Gender  HourlyRate  JobInvolvement  JobLevel JobRole  JobSatisfaction  \
1470    NaN         NaN             NaN       NaN     NaN              NaN   
1471    NaN         NaN             NaN       NaN     NaN              NaN   
1472    NaN         NaN             NaN       NaN     NaN              NaN   
1473    NaN         NaN             NaN       NaN     NaN              NaN   
1474    NaN         NaN             NaN       NaN     NaN              NaN   
1475    NaN         NaN             NaN       NaN     NaN              NaN   

     MaritalStatus  MonthlyIncome  MonthlyRate  NumCompaniesWorked Over18  \
1470           NaN            NaN          NaN                 NaN    NaN   
1471           NaN            NaN          NaN                 NaN    NaN   
1472           NaN            NaN          NaN                 NaN    NaN   
1473           NaN            NaN          NaN                 NaN    NaN   
1474           NaN            NaN          NaN                 NaN    NaN   
1475           NaN            NaN          NaN                 NaN    NaN   

     OverTime  PercentSalaryHike  PerformanceRating  RelationshipSatisfaction  \
1470      NaN                NaN                NaN                       NaN   
1471      NaN                NaN                NaN                       NaN   
1472      NaN                NaN                NaN                       NaN   
1473      NaN                NaN                NaN                       NaN   
1474      NaN                NaN                NaN                       NaN   
1475      NaN                NaN                NaN                       NaN   

      StandardHours  StockOptionLevel  TotalWorkingYears  \
1470            NaN               NaN                NaN   
1471            NaN               NaN                NaN   
1472            NaN               NaN                NaN   
1473            NaN               NaN                NaN   
1474            NaN               NaN                NaN   
1475            NaN               NaN                NaN   

      TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
1470                    NaN              NaN             NaN   
1471                    NaN              NaN             NaN   
1472                    NaN              NaN             NaN   
1473                    NaN              NaN             NaN   
1474                    NaN              NaN             NaN   
1475                    NaN        

In [70]:
# Will drop these rows
df = df.dropna(how='any')
print(f'New shape: {df.shape}')

New shape: (1470, 35)


In [71]:
# Verify no missing values
df.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince

Identified 6 rows with missing values across almost all columns. These rows appeared to be data export artificacts and comprised less than 1% so they were removed from the dataset. The dataset now contains 1470 employees with no missing values.

### Duplicate Rows
Check for any rows that are duplicates.

In [72]:
# Check for duplicates
print(f'Number of duplicate rows: {df.duplicated().sum()}')

Number of duplicate rows: 0


There no duplicate rows.

### Data Types 

In [73]:
# Check data types of variables
df.dtypes

Age                          object
Attrition                    object
BusinessTravel               object
DailyRate                   float64
Department                   object
DistanceFromHome            float64
Education                   float64
EducationField               object
EmployeeCount               float64
EmployeeNumber              float64
EnvironmentSatisfaction     float64
Gender                       object
HourlyRate                  float64
JobInvolvement              float64
JobLevel                    float64
JobRole                      object
JobSatisfaction             float64
MaritalStatus                object
MonthlyIncome               float64
MonthlyRate                 float64
NumCompaniesWorked          float64
Over18                       object
OverTime                     object
PercentSalaryHike           float64
PerformanceRating           float64
RelationshipSatisfaction    float64
StandardHours               float64
StockOptionLevel            

In [74]:
# Look at categorical variables
cat_cols = df.select_dtypes(include = ['object']).columns
print(f'Categorical columns: {list(cat_cols)}')

Categorical columns: ['Age', 'Attrition', 'BusinessTravel', 'Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus', 'Over18', 'OverTime']


In [75]:
# Unique values for categorical variables
for col in cat_cols:
    print(f'\n{col}: {df[col].nunique()} unique values')
    print(df[col].value_counts())


Age: 43 unique values
Age
35    78
34    77
36    69
31    69
29    68
32    61
30    60
33    58
38    58
40    57
37    50
27    48
28    48
42    46
39    42
45    41
41    40
26    39
44    33
46    33
43    32
50    30
25    26
24    26
49    24
47    24
55    22
51    19
53    19
48    19
54    18
52    18
22    16
56    14
23    14
58    14
21    13
20    11
59    10
19     9
18     8
60     5
57     4
Name: count, dtype: int64

Attrition: 2 unique values
Attrition
No     1233
Yes     237
Name: count, dtype: int64

BusinessTravel: 3 unique values
BusinessTravel
Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: count, dtype: int64

Department: 3 unique values
Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64

EducationField: 6 unique values
EducationField
Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other               

In [76]:
# Convert Age to numeric
df['Age'] = pd.to_numeric(df['Age'], errors='coerce')
# Check for NaNs
df['Age'].isnull().sum()

np.int64(0)

In [77]:
# Remove Over18 column - it is constant feature
print(df['Over18'].nunique())
df = df.drop('Over18', axis=1)
print(f'New shape: {df.shape}')

1
New shape: (1470, 34)


In [78]:
# Get numeric columns
num_cols = df.select_dtypes(include = ['float64', 'int64']).columns
num_cols

Index(['Age', 'DailyRate', 'DistanceFromHome', 'Education', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'MonthlyIncome',
       'MonthlyRate', 'NumCompaniesWorked', 'PercentSalaryHike',
       'PerformanceRating', 'RelationshipSatisfaction', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole',
       'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')

In [79]:
# Check for low cardinality - may intake variable should be categorical
for col in num_cols:
    n_unique = df[col].nunique()
    print(f'{col}: {n_unique} unique values')
    if n_unique <=10:
        print(f' Values: {sorted(df[col].unique())}')
    print()

Age: 43 unique values

DailyRate: 886 unique values

DistanceFromHome: 29 unique values

Education: 5 unique values
 Values: [np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0), np.float64(5.0)]

EmployeeCount: 1 unique values
 Values: [np.float64(1.0)]

EmployeeNumber: 1470 unique values

EnvironmentSatisfaction: 4 unique values
 Values: [np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0)]

HourlyRate: 71 unique values

JobInvolvement: 4 unique values
 Values: [np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0)]

JobLevel: 5 unique values
 Values: [np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0), np.float64(5.0)]

JobSatisfaction: 4 unique values
 Values: [np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0)]

MonthlyIncome: 1349 unique values

MonthlyRate: 1427 unique values

NumCompaniesWorked: 10 unique values
 Values: [np.float64(0.0), np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0)

In [80]:
# Remove EmployeeCount and StandardHours - have constant value
df = df.drop(['EmployeeCount', 'StandardHours'], axis=1)
print(f'New shape: {df.shape}')

New shape: (1470, 32)


In [81]:
# Remove EmployeeNumber as it is only an ID number without predictive value
df = df.drop('EmployeeNumber', axis=1)

In [82]:
# Print categorical and numeric variables
cat_cols = df.select_dtypes(include = ['object']).columns
print(f'Number of categorical variables: {len(cat_cols)}')
print(f'List of categorical variables: {cat_cols}')

num_cols = [c for c in df if c not in cat_cols]
print(f'\nNumber of numeric variables: {len(num_cols)}')
print(f'List of numeric variables: {num_cols}')

Number of categorical variables: 8
List of categorical variables: Index(['Attrition', 'BusinessTravel', 'Department', 'EducationField', 'Gender',
       'JobRole', 'MaritalStatus', 'OverTime'],
      dtype='object')

Number of numeric variables: 23
List of numeric variables: ['Age', 'DailyRate', 'DistanceFromHome', 'Education', 'EnvironmentSatisfaction', 'HourlyRate', 'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked', 'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager']


**Data Type Adjustments**

EmployeeCount, StandardHours, and Over18: Removed due to constant value. EmployeeNumber removed as ID only and no predictive value.

Ordinal Variables: Several columns (Education, EnvironmentSatisfaction, JobInvolvement, JobLevel, JobSatisfaction, RelationshipSatisfaction, StockOptionLevel, WorkLifeBalance) contain ordinal categorical data encoded as integers. These represent ordered ratings/levels and will be kept as numeric for modeling, as the ordering contains meaningful information.

Age: Converted to numeric type.

### Target Variable
Attrition: Binary Indicator Variable

In [83]:
# Evaluate target variable
df['Attrition'].value_counts()

Attrition
No     1233
Yes     237
Name: count, dtype: int64

In [84]:
# Percentage for attrition
df['Attrition'].value_counts(normalize=True)*100

Attrition
No     83.877551
Yes    16.122449
Name: proportion, dtype: float64

In [85]:
# Convert 0/1 binary for modeling
df['Attrition'] = df['Attrition'].map({'Yes':1, 'No':0})
df['Attrition'].value_counts()

Attrition
0    1233
1     237
Name: count, dtype: int64

In [86]:
# Change attrition data type to object
df['Attrition'] = df['Attrition'].astype('object')

**Target Variable Analysis**

The target variable, **Attrition**, shows class imbalance:
- **No**: 83.88% (1233 employees)
- **Yes**: 16.12% (237 employees)

There is a moderate class imbalance. Techniques to consider:
- SMOTE
- Class weights in model training
- Stratifiedt sampling in train-test split

Converted to binary indicator 0/1. 0 = No, 1 = Yes.

### Summary
The cleaned dataset has 1470 employees with 32 variables. Variables include the target outcome (0/1 for attrition) and demographic, job-related, and salary information. 

There were 6 rows with missing values, which were removed as they did not contain actual employee information and comprised less than 1% of data. No duplicate rows were found.

Age was converted to numeric data type. EmployeeCount, StandardHours, and Over18 were removed for constant values. EmployeeNumber removed as ID only.

The target variable shows moderate class imbalance, which will be addressed during modeling.

The dataset is now ready for exploratory data analysis.

In [87]:
# Save cleaned dataset
df.to_csv('../data/HR-Employee-Attrition-Cleaned.csv', index=False)